### PR TITLE
Revival

### PR DESCRIPTION
**Pull request by h-vetinari**
Saw this feedstock didn't have an update in a while - it's going to become a dependency for [allennlp 2.0.0](https://github.com/allenai/allennlp/blob/v2.0.0rc1/setup.py#L54), and I'm co-maintaining that [feedstock](https://github.com/conda-forge/allennlp-feedstock)

Closes #9
Closes #11 

Possible follow-ups:
* build with cuda
* build with ffmpeg support
* turn into multi-output recipe (e.g. with/without ffmpeg/image support)?
* package c++ lib separately?
* build for win/osx

### CODE REVIEW

**Comment by h-vetinari** [comment]
@isuruf @hmaarrfk @nehaljwani 
This builds, runs the test-suite, and passes except for one py37-only failure (that I don't understand). Not sure if we want to have ffmpeg & cuda support already in this PR or as follow-ups, but I think this is pretty close.

**Comment by h-vetinari** [comment]
The failure on 3.8 is flaky (was green before), and is the same as on 3.7. I commented in an issue upstream, let's see if there's a response.

**Comment by h-vetinari** [comment]
Ping @isuruf @hmaarrfk @nehaljwani

I'd be happy to skip the flaky test for now - the current status of this feedstock was without the test suite anyway (imports only), so this is already a big improvement.

WDYT?

**Comment by isuruf** [comment]
I think we should have gpu builds before merging this. It would be a surprise if `pytorch` was built with cuda and this wasn't.

Yes, we can skip that test.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -7,36 +10,89 @@ package:
...
+        # split off image/video into separate outputs?
+        - jpeg
+        - libpng
+        # ffmpeg
+        - numpy
+        - pillow >=4.1.1
+        - pytorch
+      run:
+        - python
+        - pytorch
```
Comment (isuruf) [inline recipe/meta.yaml:61]: We need to restrict this to `pytorch` cpu builds for `torchvision` cpu builds and `pytorch` gpu builds for `torchvision` gpu builds.

**Comment by h-vetinari** [comment]
The `typing_extensions` stuff seems to be breaking the python 3.6/3.7 builds, but 3.8 is green. :)
```
ModuleNotFoundError: No module named 'typing_extensions'
```

**Comment by h-vetinari** [comment]
Ah, that's probably because https://github.com/conda-forge/pytorch-cpu-feedstock/commit/41469c263557d3ab9998060b9034db19eaa095e1 ran with `ci skip`...

**Comment by h-vetinari** [comment]
@isuruf
This is green - should be ready (at least for review).

File: recipe/meta.yaml [hunk 2]
```diff
@@ -7,36 +10,93 @@ package:
...
+      source_files:
+        - test
+      imports:
+        - torchvision
+        - torchvision.datasets
+        - torchvision.models
+        - torchvision.transforms
+        - torchvision.utils
+      commands:
+        - pytest test/
```
Comment (h-vetinari) [inline recipe/meta.yaml:83]: Although, I can't see these tests being run anywhere.

Comment (h-vetinari) [reply recipe/meta.yaml:83]: Might be https://github.com/conda/conda-build/issues/4172 again.

**Comment by h-vetinari** [comment]
```
        torch._C._cuda_init()
    RuntimeError: No CUDA GPUs are available
```

Looks like the way upstream wants to build has runtime dependency on CUDA hardware. We coud probably patch that out?

File: recipe/meta.yaml [hunk 3]
```diff
@@ -56,12 +56,16 @@ outputs:
         - numpy
         - pillow >=4.1.1
         - pytorch =*={{ torch_proc_type }}*
+        # TODO: remove this once it is properly represented in pytorch build metadata
+        - typing_extensions  # [py<38]
```
Comment (h-vetinari) [inline recipe/meta.yaml]: Has been swiftly fixed by https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/118

**Comment by h-vetinari** [comment]
So my idea of patching this is not possible in this feedstock, since failing check is in `CUDAExtension`, which is not part of `torchvision` but comes from `pytorch` [directly](https://github.com/pytorch/vision/tree/v0.8.2#L13):
```
from torch.utils.cpp_extension import BuildExtension, CppExtension, CUDAExtension, CUDA_HOME
```

**Comment by hmaarrfk** [comment]
do you want to monkey patch?

**Comment by h-vetinari** [comment]
> 
> 
> do you want to monkey patch?

Not sure what other options we'd have. Maybe @isuruf has some ideas?

**Comment by hmaarrfk** [comment]
@jakirkham do you have any ideas on how to get around this? torch tries to determine gpu capabilities at build time.

**Comment by dirkgr** [comment]
You probably already know this, but torch itself recommends installing `torchvision` with `conda install pytorch torchvision -c pytorch` or `conda install pytorch torchvision cudatoolkit=11.0 -c pytorch`. Their version works with and without GPUs, so there must be a way to compile it one way that works for both.

**Comment by h-vetinari** [comment]
> 
> 
> You probably already know this, but torch itself recommends installing `torchvision` with `conda install pytorch torchvision -c pytorch` or `conda install pytorch torchvision cudatoolkit=11.0 -c pytorch`. Their version works with and without GPUs, so there must be a way to compile it one way that works for both.

It's certainly possible to compile both in principle. What's not possible is using packages from outside conda-forge to build packages for conda-forge (with the sole exception of the anaconda default channels). Maybe we can get some inspiration from the pytorch recipes, provided they're licensed appropriately.

**Comment by jakirkham** [comment]
> do you have any ideas on how to get around this? torch tries to determine gpu capabilities at build time.

No idea. Would ask the PyTorch developers for guidance

**Comment by isuruf** [comment]
@conda-forge-admin, rerender

**Comment by isuruf** [comment]
We can restart cuda 11 builds when pytorch cuda 11 builds are uploaded.

**Comment by h-vetinari** [comment]
Thanks for finishing this off, @isuruf!

File: recipe/build-torch.sh [hunk 4]
```diff
@@ -1,6 +1,21 @@
...
+  elif [[ ${cuda_compiler_version} == 10.* ]]; then
+      export TORCH_CUDA_ARCH_LIST="$TORCH_CUDA_ARCH_LIST;6.0;6.1;7.0;7.5"
+  elif [[ ${cuda_compiler_version} == 11.0* ]]; then
+      export TORCH_CUDA_ARCH_LIST="$TORCH_CUDA_ARCH_LIST;6.0;6.1;7.0;7.5;8.0"
+  elif [[ ${cuda_compiler_version} == 11.1* ]]; then
+      export TORCH_CUDA_ARCH_LIST="$TORCH_CUDA_ARCH_LIST;6.0;6.1;7.0;7.5;8.0;8.6"
+  else
+      echo "unsupported cuda version. edit build.sh"
+      exit 1
+  fi
```
Comment (h-vetinari) [inline recipe/build-torch.sh:18]: @isuruf
Shouldn't that `+PTX` be on the last available architecture?

File: recipe/meta.yaml [hunk 5]
```diff
@@ -1,42 +1,98 @@
...
+    # skip failing test thats already marked as flaky upstream, see
+    # https://github.com/pytorch/vision/blob/v0.8.2/test/test_models.py#L90
+    - patches/0001-skip-flaky-test.patch
 
 build:
-  number: 1000
-  skip: True  # [not linux64]
-  script: python setup.py install --single-version-externally-managed --record=record.txt
+  number: {{ number }}
+  skip: true  # [not linux]
```
Comment (h-vetinari) [inline recipe/meta.yaml:21]: Setting an anchor for easy referencing here...

**Pull request merged**